server/service/transport_error.go

```diff
@@ -33,66 +33,75 @@ func baseError(err string) []map[string]string {
 	}
 }
 
+type validationErrorInterface interface {
+	error
+	Invalid() []map[string]string
+}
+
+type permissionErrorInterface interface {
+	PermissionError() []map[string]string
+}
+
+type notFoundErrorInterface interface {
+	error
+	IsNotFound() bool
+}
+
+type existsErrorInterface interface {
+	error
+	IsExists() bool
+}
+
+type causerInterface interface {
+	Cause() error
+}
+
 // encode error and status header to the client
 func encodeError(ctx context.Context, err error, w http.ResponseWriter) {
 	enc := json.NewEncoder(w)
 	enc.SetIndent("", "  ")
 
-	type validationError interface {
-		error
-		Invalid() []map[string]string
-	}
-	if e, ok := err.(validationError); ok {
+	err = getRootCause(err)
```

Comment:
I think `errors.Cause(err)` does the equivalent of this (https://pkg.go.dev/github.com/pkg/errors?utm_source=godoc#Cause)

Reply:
Ah, yes, I misread the `for err != nil {` as an if. Thanks for catching that!